File: www/libs/coweb/jsoe/OperationEngine.js
//
// Operation engine public API.
//
// Copyright (c) The Dojo Foundation 2011. All Rights Reserved.
// Copyright (c) IBM Corporation 2008, 2011. All Rights Reserved.
//
dojo.provide('coweb.jsoe.OperationEngine');
dojo.require('coweb.jsoe.ContextVectorTable');
dojo.require('coweb.jsoe.ContextVector');
dojo.require('coweb.jsoe.HistoryBuffer');
dojo.require('coweb.jsoe.Operation');

/**
 * Controls the operational transformation algorithm. Provides a public
 * API for operation processing, garbage collection, engine synchronization.
 *
 * @ivar siteId Unique site integer for this engine instance in a conference
 * @ivar cv Context vector representing local document state
 * @ivar cvt Context vector table representing all conference document states
 * @ivar hb History buffer of local and remote operations
 * @ivar typeMap Mapping from string op names to op subclasses
 */
dojo.declare('coweb.jsoe.OperationEngine', null, {
    /**
     * Initializes the engine components.
     * 
     * @param site Local site integer to assign to this engine instance
     */
    constructor: function(site) {
        this.siteId = site;
        this.cv = new coweb.jsoe.ContextVector({count : site+1});
        this.cvt = new coweb.jsoe.ContextVectorTable(this.cv, site);
        this.hb = new coweb.jsoe.HistoryBuffer();
    },

    /**
     * Gets the state of this engine instance to seed a new instance.
     *
     * @return Array of cvt state, hb state, local site ID, and indices of
     *   frozen cvt entries
     */
    getState: function() {
        // op engine state can be cloned from cvt, hb, site ID, and frozen slots
        // get indices of frozen cvt slots
        var frozen = this.cvt.getEquivalents(this.cv, this.siteId);
        return [this.cvt.getState(), this.hb.getState(), this.siteId, frozen];
    },

    /**
     * Sets the state of this engine instance to state received from another
     * instance in a conference.
     *
     * @param arr Array of values returned by getState on a remote instance
     */
    setState: function(arr) {
        // configure the history buffer and context vector table
        this.cvt.setState(arr[0]);
        this.hb.setState(arr[1]);
        // pull out the context vector for the sending site
        this.cv = this.cvt.getContextVector(arr[2]);
        // copy it
        this.cv = this.cv.copy();
        // freeze our own site slot
        this.cvt.updateWithContextVector(this.siteId, this.cv);
        // freeze all sites that must be frozen
        var frozen = arr[3];
        for(var i=0; i < frozen.length; i++) {
            this.freezeSite(frozen[i]);
        }
    },

    /**
     * Makes a copy of the engine context vector representing document state.
     *
     * @return Context vector
     */
    copyContextVector: function() {
        return this.cv.copy();
    },

    /**
     * Factory method that creates an operation object initialized with the
     * given values.
     *
     * @param local True if the operation was originated locally, false if not
     * @param key String key identifying what property this op affects
     * @param value Arbitrary new data value associated with the property
     * @param type String type of operation: update, insert, delete
     * @param position Integer position of the property in a linear order
     * @param site Site integer on which the op originated (remote only)
     * @param cv Context vector timestamp for the op (remote only)
     * @return Operation subclass instance matching the given type
     */
    createOp: function(local, key, value, type, position, site, cv) {
        var args;
        if(local) {
            args = {
                key : key,
                position : position,
                value : value,
                siteId : this.siteId,
                contextVector : this.copyContextVector()
            };
        } else {
            // build cv from raw sites array
            cv = new coweb.jsoe.ContextVector({sites : cv});
            args = {
                key : key,
                position : position,
                value : value,
                siteId : site,
                contextVector : cv
            };
        }
        return coweb.jsoe.Operation.createFromType(type, args);
    },

    /**
     * Creates an operation object and pushes it into the operation engine
     * algorithm.
     *
     * @param local True if the operation was originated locally, false if not
     * @param key String key identifying what property this op affects
     * @param value Arbitrary new data value associated with the op
     * @param type String type of operation: update, insert, delete
     * @param position Integer position of the property in a linear order
     * @param site Site integer on which the op originated (remote only)
     * @param cv Context vector timestamp for the op (remote only)
     * @return Original operation if local, transformed operation if remote
     */
    push: function(local, key, value, type, position, site, cv) {
        var op = this.createOp(local, key, value, type, position, site, cv);
        if(local) {
            return this.pushLocalOp(op);
        } else {
            return this.pushRemoteOp(op);
        }
    },

    /**
     * Procceses a local operation.
     *
     * @param op Operation object
     * @return Operation object (same ref as param)
     */
    pushLocalOp: function(op) {
        // update local context vector
        this.cv.setSeqForSite(op.siteId, op.seqId);
        // add to history buffer
        this.hb.add(op);
        return op;
    },

    /**
     * Procceses a remote operation.
     *
     * @param op Operation object
     * @return Transformed operation object (not same ref as param)
     */
    pushRemoteOp: function(op) {
        var top = null;

        if(this.hasProcessedOp(op)) {
            // engine has already processed this op so ignore it
            return null;
        } else if(this.cv.equals(op.contextVector)) {
            // no transform needed
            // make a copy so return value is independent of input
            top = op.copy();
        } else {
            // transform needed to upgrade context
            var cd = this.cv.subtract(op.contextVector);
            // make the original op immutable
            op.immutable = true;
            // top is a transformed copy of the original
            top = this._transform(op, cd);
        }

        // update local context vector with the original op
        this.cv.setSeqForSite(op.siteId, op.seqId);
        // store original op
        this.hb.add(op);
        // update context vector table with original op
        this.cvt.updateWithOperation(op);

        // return the transformed op
        return top;
    },

    /**
     * Processes an engine synchronization event.
     *
     * @param site Integer site ID of where the sync originated
     * @param cv Context vector of that site
     */
    pushSync: function(site, cv) {
        // update the context vector table
        this.cvt.updateWithContextVector(site, cv);
    },

    /**
     * Processes an engine synchronization event.
     *
     * @param site Integer site ID of where the sync originated
     * @param site Array of site sequence values for a context vector
     */
    pushSyncWithSites: function(site, sites) {
        // build a context vector from raw site data
        var cv = new coweb.jsoe.ContextVector({state : sites});
        this.pushSync(site, cv);
    },

    /**
     * Runs the garbage collection algorithm over the history buffer.
     *
     * @return Minimum context vector object or null if gc didn't run
     */
    purge: function() {
        if(this.getBufferSize() === 0) {
            // exit quickly if there is nothing to purge
            return null;
        }
        // get minimum context vector
        var mcv = this.cvt.getMinimumContextVector();
        
        if(mcv === null) {
            // exit quickly if there is no mcv
            return null;
        }

        var min_op; 
        var cd = this.cv.oldestDifference(mcv);
        var ops = this.hb.getOpsForDifference(cd);
        while(ops.length) {
            // get an op from the list we have yet to process
            var curr = ops.pop();
            // if we haven't pick a minimum op yet OR
            // the current op is sorted before minimum op in the hb
            if(min_op === undefined || curr.compare(min_op) == -1) {
                // compute the oldest difference between the document state
                // and the current op
                cd = this.cv.oldestDifference(curr.contextVector);
                // add the operations in this difference to the list to process
                ops = ops.concat(this.hb.getOpsForDifference(cd));
                // make the current op the new minimum
                min_op = curr;
            }
        }

        // get history buffer contents sorted by context dependencies
        ops = this.hb.getSortedOperations();
        // remove keys until we hit the min
        for(var i=0; i < ops.length; i++) {
            var op = ops[i];
            // if there is no minimum op OR
            // if this op is not the minimium
            if(min_op === undefined || 
               (min_op.siteId != op.siteId || min_op.seqId != op.seqId)) {
                // remove operation from history buffer
                this.hb.remove(op);
            } else {
                // don't remove any more ops with context greater than or 
                // equal to the minimum
                break;
            }
        }
        return mcv;
    },

    /**
     * Gets the size of the history buffer in terms of stored operations.
     * 
     * @return Integer size
     */
    getBufferSize: function() {
        return this.hb.getCount();
    },

    /**
     * Gets if the engine has already processed the give operation based on
     * its context vector and the context vector of this engine instance.
     *
     * @param Operation object
     * @return True if already processed, false if not
     */
    hasProcessedOp: function(op) {
        var seqId = this.cv.getSeqForSite(op.siteId);
        return (seqId >= op.seqId);
    },

    /**
     * Freezes a slot in the context vector table by inserting a reference
     * to this engine instance's context vector.
     *
     * @param site Integer ID of the site to freeze
     */
    freezeSite: function(site) {
        // insert a ref to this site's cv into the cvt for the given site
        this.cvt.updateWithContextVector(site, this.cv);
    },

    /**
     * Thaws a slot in the context vector table by inserting an empty context
     * vector into the context vector table.
     *
     * @param site Integer ID of the site to thaw
     */
    thawSite: function(site) {
        // don't ever thaw the slot for our own site
        if(site == this.siteId) {return;}
        // get the minimum context vector
        var cv = this.cvt.getMinimumContextVector();
        // grow it to include the site if needed
        cv.growTo(site);
        // use it as the initial context of the site
        this.cvt.updateWithContextVector(site, cv);
    },

    /**
     * Executes a recursive step in the integration algorithm.
     *
     * @param op Operation to transform
     * @param cd Context vector difference between the given op and another
     *   document state
     * @return Transformed operation (not a ref to the param op)
     */
    _transform: function(op, cd) {
        // get all ops for context different from history buffer sorted by
        //   context dependencies
        var ops = this.hb.getOpsForDifference(cd);
        // copy the incoming operation to avoid disturbing the history buffer
        //   when the op comes from our history buffer during a recursive step
        op = op.copy();
        //iterate over all operations in the difference
        for(var i=0; i < ops.length; i++) {
            // xop is the previously applied op
            var xop = ops[i];
            if(!op.contextVector.equals(xop.contextVector)) {
                // transform needed to upgrade context of xop to op
                var xcd = op.contextVector.subtract(xop.contextVector);
/*                if(!xcd.sites.length) {
                    throw new Error('bad context diff');
                }*/
                // we'll get a copy back from the recursion
                var cxop = this._transform(xop, xcd);
                if(cxop === null) {
                    // xop was invalidated by a previous op during the transform
                    // so it has no effect on the current op; upgrade context
                    // immediately
                    op.contextVector.setSeqForSite(xop.siteId, xop.seqId);
                    continue;
                }
                // now only deal with the copy
                xop = cxop;
            }
/*            if(!op.contextVector.equals(xop.contextVector)) {
                throw new Error('context vectors unequal after upgrade');
            }*/
            // perform the inclusion transform on op and xop now that they have
            //   the same context; ask xop for the method that should be invoked
            //   on op to properly transform it
            op = op[xop.transformMethod()](xop);
            if(op === null) {
                // op target was deleted by another earlier op so return now
                // do not continue because no further transforms have any
                // meaning on this op
                return null;
            }
            // upgrade the context of the transformed op to reflect the
            //   inclusion transform
            op.contextVector.setSeqForSite(xop.siteId, xop.seqId);
        }
        // op is always a copy because we never entered this method if no
        // transform was needed
        return op;
    }
});

File: servers/java/coweb-operationengine/src/main/java/org/coweb/oe/OperationEngine.java
package org.coweb.oe;

import java.util.Map;
import java.util.HashMap;
import java.util.Stack;

public class OperationEngine {

	private int siteId;
	private ContextVector cv = null;
	private ContextVectorTable cvt = null;
	private HistoryBuffer hb = null;
	private int siteCount = 1;

	/**
     * Controls the operational transformation algorithm. Provides a public
     * API for operation processing, garbage collection, and engine 
     * synchronization.
     *
     * @constructor
     * @param {Number} siteId Unique integer site ID for this engine instance
     */
	public OperationEngine(int siteId) throws OperationEngineException {
		this.siteId = siteId;

		HashMap<String, Object>args = new HashMap<String, Object>();
		args.put("count", siteId + 1);
		this.cv = new ContextVector(args);
		this.cvt = new ContextVectorTable(this.cv, siteId);
		this.hb = new HistoryBuffer();
	}

	/**
     * Gets the state of this engine instance to seed a new instance.
     *
     * @return {Object[]} Array or serialized state
     */
	public Object[] getState() {

		int[] frozen = this.cvt.getEquivalents(this.cv,
				this.siteId);

		Object[] ret = { this.cvt.getState(), this.hb.getState(), new Integer(this.siteId),
				frozen };

		return ret;
	}

	/**
     * Sets the state of this engine instance to state received from another
     * instance.
     *
     * @param {Object[]} arr Array in the format returned by getState
     */
	public void setState(Object[] arr) throws OperationEngineException {
		this.cvt.setState((int[][])arr[0]);
		this.hb.setState((Object[])arr[1]);

		this.cv = this.cvt.getContextVector(((Integer)arr[2]).intValue());

		this.cv = this.cv.copy();

		this.cvt.updateWithContextVector(this.siteId, this.cv);

		this.siteCount = this.cv.getSize();

		int[] frozen = (int[]) arr[3];

		for (int i = 0; i < frozen.length; i++) {
			this.freezeSite(frozen[i]);
		}
	}

	/**
	 * Makes a copy of the engine context vector representing the local document
	 * state.
	 * @throws OperationEngineException 
	 * 
	 * @returns {ContextVector} Copy of the context vector for the local site
	 */
	public ContextVector copyContextVector() throws OperationEngineException {
		return this.cv.copy();
	}

	/**
	 * Factory method that creates an operation object initialized with the
	 * given values.
	 * 
	 * @param {Boolean} local True if the operation was originated locally,
	 *        false if not
	 * @param {String} key Operation key
	 * @param {String} value Operation value
	 * @param {String} type Type of operation: update, insert, delete
	 * @param {Number} position Operation integer position
	 * @param {Number} site Integer site ID where a remote op originated.
	 *        Ignored for local operations which adopt the local site ID.
	 * @param {ContextVector} cv Operation context. Ignored for local operations
	 *        which adopt the local site context.
	 * @param {Number} order Place of the operation in the total order. Ignored
	 *        for local operations which are not yet assigned a place in the
	 *        order.
	 * @throws OperationEngineException 
	 * @returns {Operation} Subclass instance matching the given type
	 */
	public Operation createOp(boolean local, String key, String value,
			String type, int position, int site, ContextVector cv, int order) throws OperationEngineException {
		Map<String, Object> args = new HashMap<String, Object>();
		if (local) {
			args.put("key", key);
			args.put("position", new Integer(position));
			args.put("value", value);
			args.put("siteId", new Integer(this.siteId));
			args.put("contextVector", this.copyContextVector());
			args.put("local", true);
		} else {
			// build cv from raw sites array
			HashMap<String, Object> map = new HashMap<String, Object>();
			map.put("sites", cv);
			cv = new ContextVector(map);

			args.put("key", key);
			args.put("position", new Integer(position));
			args.put("value", value);
			args.put("siteId", new Integer(site));
			args.put("contextVector", cv);
			args.put("order", order);
			args.put("local", false);
		}

		return Operation.createOperationFromType(type, args);
	}

	/**
	 * Creates an operation object and pushes it into the operation engine
	 * algorithm. The parameters and return value are the same as those
	 * documented for createOp.
	 * @throws OperationEngineException 
	 */
	public Operation push(boolean local, String key, String value, String type,
			int position, int site, ContextVector cv, int order) throws OperationEngineException {

		Operation op = this.createOp(local, key, value, type, position, site,
				cv, order);
		if (local) {
			return this.pushLocalOp(op);
		} else {
			return this.pushRemoteOp(op);
		}
	}

	/**
	 * Procceses a local operation and adds it to the history buffer.
	 * 
	 * @param {Operation} Local operation
	 * @returns {Operation} Reference to the pass parameter
	 */
	public Operation pushLocalOp(Operation op) {
		// update local context vector
		this.cv.setSeqForSite(op.getSiteId(), op.getSeqId());
		// add to history buffer
		this.hb.addLocal(op);
		return op;
	}

	/**
	 * Procceses a remote operation, transforming it if required, and adds the
	 * original to the history buffer.
	 * 
	 * @param {Operation} Remote operation
	 * @throws OperationEngineException 
	 * @returns {Operation|null} New, transformed operation object or null if
	 *          the effect of the passed operation is nothing and should not be
	 *          applied to the shared state
	 */
	public Operation pushRemoteOp(Operation op) throws OperationEngineException {
		Operation top = null;

		if (this.hasProcessedOp(op)) {
			// let the history buffer track the total order for the op
			this.hb.addRemote(op);
			// engine has already processed this op so ignore it
			return null;
		} else if (this.cv.equals(op.getContextVector())) {
			// no transform needed
			// make a copy so return value is independent of input
			top = op.copy();
		} else {
			// transform needed to upgrade context
			ContextDifference cd = this.cv.subtract(op.getContextVector());
			// make the original op immutable
			op.setImmutable(true);
			// top is a transformed copy of the original
			top = this._transform(op, cd);
		}

		// update local context vector with the original op
		this.cv.setSeqForSite(op.getSiteId(), op.getSeqId());
		// store original op
		this.hb.addRemote(op);
		// update context vector table with original op
		this.cvt.updateWithOperation(op);

		// return the transformed op
		return top;
	}

	/**
	 * Processes an engine synchronization event.
	 * 
	 * @param {Number} site Integer site ID of where the sync originated
	 * @param {ContextVector} cv Context vector sent by the engine at that site
	 * @throws OperationEngineException 
	 */
	public void pushSync(int site, ContextVector cv) throws OperationEngineException {
		// update the context vector table
		this.cvt.updateWithContextVector(site, cv);
	}

	/**
	 * Processes an engine synchronization event.
	 * 
	 * @param {Number} site Integer site ID of where the sync originated
	 * @param {Number[]} Array form of the context vector sent by the site
	 * @throws OperationEngineException 
	 */
	public void pushSyncWithSites(int site, int[] sites) throws OperationEngineException {
		// build a context vector from raw site data
		HashMap<String, Object> args = new HashMap<String, Object>();
		args.put("sites",sites);
		ContextVector cv = new ContextVector(args);
		this.pushSync(site, cv);
	}

	/**
	 * Runs the garbage collection algorithm over the history buffer.
	 * @throws OperationEngineException 
	 * 
	 * @returns {ContextVector|null} Compiuted minimum context vector of the
	 *          earliest operation garbage collected or null if garbage
	 *          collection did not run
	 */
	public ContextVector purge() throws OperationEngineException {
		if (this.getBufferSize() == 0) {
			// exit quickly if there is nothing to purge
			return null;
		}
		// get minimum context vector
		ContextVector mcv = this.cvt.getMinimumContextVector();

		if (mcv == null) {
			// exit quickly if there is no mcv
			return null;
		}

		Operation min_op = null;
		ContextDifference cd = this.cv.oldestDifference(mcv);
		Stack<Operation> ops = this.hb.getOpsForDifference(cd);
		while (ops.size() > 0) {
			// get an op from the list we have yet to process
			Operation curr = ops.pop();
			// if we haven't picked a minimum op yet OR
			// the current op is before the minimum op in context
			if (min_op == null || curr.compareByContext(min_op) == -1) {
				// compute the oldest difference between the document state
				// and the current op
				cd = this.cv.oldestDifference(curr.getContextVector());
				// add the operations in this difference to the list to process
				ops.addAll(this.hb.getOpsForDifference(cd));
				// make the current op the new minimum
				min_op = curr;
			}
		}

		// get history buffer contents sorted by context dependencies
		ops = this.hb.getContextSortedOperations();
		// remove keys until we hit the min
		for (int i = 0; i < ops.size(); i++) {
			Operation op = ops.elementAt(i);
			// if there is no minimum op OR
			// if this op is not the minimium
			if (min_op == null
					|| (min_op.getSiteId() != op.getSiteId() || min_op
							.getSeqId() != op.getSeqId())) {
				// remove operation from history buffer
				this.hb.remove(op);
			} else {
				// don't remove any more ops with context greater than or
				// equal to the minimum
				break;
			}
		}
		return mcv;
	}

	/**
	 * Gets the size of the history buffer in terms of stored operations.
	 * 
	 * @returns {Number} Integer size
	 */
	public int getBufferSize() {
		return this.hb.getCount();
	}

	/**
	 * Gets if the engine has already processed the give operation based on its
	 * context vector and the context vector of this engine instance.
	 * 
	 * @param {Operation} op Operation to check
	 * @returns {Boolean} True if the engine already processed this operation,
	 *          false if not
	 */
	public boolean hasProcessedOp(Operation op) {
		int seqId = this.cv.getSeqForSite(op.getSiteId());
		// console.log('op processed? %s: this.cv=%s, seqId=%d, op.siteId=%d,
		// op.cv=%s, op.seqId=%d',
		// (seqId >= op.seqId), this.cv.toString(), seqId, op.siteId,
		// op.contextVector.toString(), op.seqId);
		return (seqId >= op.getSeqId());
	}

	/**
	 * Freezes a slot in the context vector table by inserting a reference to
	 * context vector of this engine. Should be invoked when a remote site stops
	 * participating.
	 * 
	 * @param {Number} site Integer ID of the site to freeze
	 * @throws OperationEngineException 
	 */
	public void freezeSite(int site) throws OperationEngineException {
		// ignore if already frozen
		if (this.cvt.getContextVector(site) != this.cv) {
			// insert a ref to this site's cv into the cvt for the given site
			this.cvt.updateWithContextVector(site, this.cv);
			// one less site participating now
			this.siteCount--;
		}
	}

	/**
	 * Thaws a slot in the context vector table by inserting a zeroed context
	 * vector into the context vector table. Should be invoked before processing
	 * the first operation from a new remote site.
	 * 
	 * @param {Number} site Integer ID of the site to thaw
	 * @throws OperationEngineException 
	 */
	public void thawSite(int site) throws OperationEngineException {
		// don't ever thaw the slot for our own site
		if (site == this.siteId) {
			return;
		}
		// get the minimum context vector
		ContextVector cv = this.cvt.getMinimumContextVector();
		// grow it to include the site if needed
		cv.growTo(site);
		// use it as the initial context of the site
		this.cvt.updateWithContextVector(site, cv);
		// one more site participating now
		this.siteCount++;
	}

	/**
	 * Gets the number of sites known to be participating, including this site.
	 * 
	 * @returns {Number} Integer count
	 */
	public int getSiteCount() {
		return this.siteCount;
	};

	/**
	 * Executes a recursive step in the operation transformation control
	 * algorithm. This method assumes it will NOT be called if no transformation
	 * is needed in order to reduce the number of operation copies needed.
	 * 
	 * @param {Operation} op Operation to transform
	 * @param {ContextDifference} cd Context vector difference between the given
	 *        op and the document state at the time of this recursive call
	 * @throws OperationEngineException 
	 * @returns {Operation|null} A new operation, including the effects of all
	 *          of the operations in the context difference or null if the
	 *          operation can have no further effect on the document state
	 */
	private Operation _transform(Operation op, ContextDifference cd) throws OperationEngineException {
		// get all ops for context different from history buffer sorted by
		// context dependencies
		Stack<Operation> ops = this.hb.getOpsForDifference(cd);
		Operation xop = null;
		ContextDifference xcd = null;
		Operation cxop = null;
		Operation cop = null;

		// xcd, xop, cxop, cop, i, l;
		// copy the incoming operation to avoid disturbing the history buffer
		// when the op comes from our history buffer during a recursive step
		op = op.copy();
		// iterate over all operations in the difference

		for (int i = 0; i < ops.size(); i++) {
			// xop is the previously applied op
			xop = ops.elementAt(i);
			if (!op.getContextVector().equals(xop.getContextVector())) {
				// see if we've cached a transform of this op in the desired
				// context to avoid recursion
				cxop = xop.getFromCache(op.getContextVector());
				// cxop = null;
				if (cxop != null) {
					xop = cxop;
				} else {
					// transform needed to upgrade context of xop to op
					xcd = op.getContextVector()
							.subtract(xop.getContextVector());
					if (xcd.sites == null || xcd.sites.size() == 0) {
						throw new OperationEngineException(
								"transform produced empty context diff");
					}
					// we'll get a copy back from the recursion
					cxop = this._transform(xop, xcd);
					if (cxop == null) {
						// xop was invalidated by a previous op during the
						// transform so it has no effect on the current op;
						// upgrade context immediately and continue with
						// the next one
						op.upgradeContextTo(xop);
						// @todo: see null below
						continue;
					}
					// now only deal with the copy
					xop = cxop;
				}
			}
			if (!op.getContextVector().equals(xop.getContextVector())) {
				throw new OperationEngineException("context vectors unequal after upgrade");
			}
			// make a copy of the op as is before transform
			cop = op.copy();
			// transform op to include xop now that contexts match IT(op, xop)
			op = op.transformWith(xop);
			if (op == null) {
				// op target was deleted by another earlier op so return now
				// do not continue because no further transforms have any
				// meaning on this op
				// @todo: i bet we want to remove this shortcut if we're
				// deep in recursion when we find a dead op; instead cache it
				// so we don't come down here again
				return null;
			}
			// cache the transformed op
			op.addToCache(this.siteCount);

			// do a symmetric transform on a copy of xop too while we're here
			xop = xop.copy();
			xop = xop.transformWith(cop);
			if (xop != null) {
				xop.addToCache(this.siteCount);
			}
		}
		// op is always a copy because we never entered this method if no
		// transform was needed
		return op;
	}
}
